presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayToString.java

```diff
@@ -0,0 +1,131 @@
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static com.facebook.presto.spi.type.TypeSignature.parseTypeSignature;
+import static com.facebook.presto.type.TypeUtils.parameterizedTypeName;
+import static com.facebook.presto.type.TypeUtils.readStructuralBlock;
+import static com.facebook.presto.util.Reflection.methodHandle;
+import static io.airlift.slice.Slices.utf8Slice;
+import static java.lang.String.format;
+
+public final class ArrayToString
+        extends ParametricScalar
+{
+    public static final ArrayToString ARRAY_TO_STRING = new ArrayToString();
+    private static final TypeSignature VARCHAR_TYPE_SIGNATURE = parseTypeSignature(StandardTypes.VARCHAR);
+    private static final String FUNCTION_NAME = "array_to_string";
+    private static final Signature SIGNATURE = new Signature(FUNCTION_NAME, ImmutableList.of(comparableTypeParameter("T")), StandardTypes.VARCHAR,
+            ImmutableList.of("array<T>", StandardTypes.VARCHAR, StandardTypes.VARCHAR),
+            false, false);
+
+    @Override
+    public Signature getSignature()
+    {
+        return SIGNATURE;
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return true;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "Concatenates the elements of the given array using a delimiter and an optional string to replace nulls";
+    }
+
+    @Override
+    public FunctionInfo specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type type = types.get("T");
+        TypeSignature valueType = type.getTypeSignature();
+        TypeSignature arrayType = parameterizedTypeName(StandardTypes.ARRAY, valueType);
+        MethodHandle methodHandle = methodHandle(ArrayToString.class, "arrayToString", Type.class, ConnectorSession.class, Slice.class, Slice.class, Slice.class);
+        Signature signature = new Signature(FUNCTION_NAME, VARCHAR_TYPE_SIGNATURE, arrayType, VARCHAR_TYPE_SIGNATURE, VARCHAR_TYPE_SIGNATURE);
+
+        return new FunctionInfo(signature, getDescription(), isHidden(), methodHandle.bindTo(type), isDeterministic(), false, ImmutableList.of(false, false, true));
+    }
+
+    public static Slice arrayToString(Type elementType, ConnectorSession session, Slice slice, Slice delimiter, Slice nullReplacement)
+    {
+        if (elementType instanceof ArrayType || elementType instanceof MapType || elementType instanceof RowType) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Unexpected element type %s", elementType.getDisplayName()));
+        }
+
+        Block arrayBlock = readStructuralBlock(slice);
+        int numElements = arrayBlock.getPositionCount();
+        DynamicSliceOutput dos = new DynamicSliceOutput(numElements * 2);
```

Comment:
we write to avoid abbreviations for variable names, so call this `sliceOutput` or `output` instead